Add tool choice AND Paralllelize BaseRecalibrator

sorelfitzgibbon opened the pull request:
# Description
### Add Tool Choice AND Parallelize BaseRecalibrator

## Add Tool Choice
This PR introduces flexible processing modes for the recalibrate-BAM pipeline, allowing users to selectively run indel realignment and/or base recalibration steps instead of always running both.

##Closes #9 
#### New Features
- **Selective Processing**: Users can now choose to run:
  - Both indel realignment and base recalibration (default behavior)
  - Indel realignment only (skip BQSR)
  - Base recalibration only (skip indel realignment, run BQSR directly on raw BAMs)

- **Parameter Validation**: Added validation to ensure at least one processing step is selected

- **Enhanced Resource Management**: Updated resource configurations for all processing modes across different compute environments (F2, F8, F16, F32)

#### Configuration Changes
**New Parameters:**
- `run_indel_realignment` (boolean, default: true) - Whether to run indel realignment
- `run_base_recalibration` (boolean, default: true) - Whether to run base recalibration

**Usage Examples:**
```
# Default: Run both steps
run_indel_realignment: true
run_base_recalibration: true

# Indel realignment only
run_indel_realignment: true
run_base_recalibration: false

# Base recalibration only  
run_indel_realignment: false
run_base_recalibration: true
```

#### Technical Changes
- **Schema Updates**: Added new boolean parameters with validation
- **Workflow Logic**: Implemented conditional processing with proper channel handling
- **Module Updates**: Enhanced base-recalibration module to support flexible input sources
- **Resource Optimization**: Comprehensive resource configurations for all compute tiers
- **Documentation**: Updated README with usage examples and parameter descriptions

#### Benefits
- **Flexibility**: Supports different analysis workflows and requirements
- **Efficiency**: Allows skipping unnecessary steps to save compute time
- **Backwards Compatible**: Default behavior unchanged (both steps run)
- **Resource Optimized**: Proper resource allocation for each processing mode


## Parallelize BaseRecalibrator Processing for Improved Performance
### Closes #94

### Overview
This PR introduces parallelization of the BaseRecalibrator step in the recalibrate-BAM pipeline, significantly improving processing performance for large datasets by leveraging interval-based parallel processing.

### Key Changes

#### 🚀 **Performance Improvements**
- **Parallelized BaseRecalibrator execution**: BaseRecalibrator now runs in parallel across genomic intervals instead of sequentially on whole genomes
- **Interval-based processing**: Each interval generates its own recalibration table, enabling concurrent processing
- **Automatic table gathering**: Individual interval tables are automatically merged using `GatherBQSRReports` 
- **Parallel BQSR application**: ApplyBQSR also runs in parallel across intervals for consistent performance gains

#### 🔧 **Technical Implementation**
- **Enhanced workflow orchestration**: Modified `module/base-recalibration.nf` to handle interval-based parallelization
- **Sample ID preservation**: Improved sample ID handling throughout the parallel processing pipeline
- **Resource optimization**: Updated `config/resources.json` with optimized memory and CPU allocations for parallel processes
- **Interval management**: Enhanced interval splitting and processing logic in `main.nf`

#### 🛠️ **Process Flow**
1. **Split intervals**: Genome is split into intervals (by chromosome or scatter count)
2. **Parallel BaseRecalibrator**: Each interval runs BaseRecalibrator independently
3. **Gather reports**: `GatherBQSRReports` combines interval-specific recalibration tables
4. **Parallel ApplyBQSR**: BQSR application runs in parallel across intervals
5. **Merge results**: Final BAM merging consolidates interval-level outputs

### Benefits
- **Faster processing**: Dramatic reduction in runtime for large BAM files through parallel processing
- **Better resource utilization**: More efficient use of available CPU cores and memory
- **Scalability**: Performance scales with available compute resources
- **Backward compatibility**: Existing configurations continue to work without modification

### Testing
- Updated test configurations (`nftest.yml`, `test/nftest.config`) to validate parallel processing
- Verified compatibility with both chromosome-based and scatter-count-based interval splitting
- Confirmed proper handling of both single and multi-sample inputs

### Configuration
The parallelization is controlled by existing parameters:
- `parallelize_by_chromosome`: Enable chromosome-based parallelization
- `scatter_count`: Number of intervals when not using chromosome-based splitting
- Resource allocations automatically optimized for parallel processing

---

This parallelization represents a significant performance improvement for the recalibrate-BAM pipeline, particularly beneficial for large-scale genomic processing


## Testing Results


- NFTest -recalibration only
  - log:  `/hot/software/pipeline/pipeline-recalibrate-BAM/Nextflow/development/1.0.2/sfitz-add-tool-choice/log-nftest-20250703T235014Z.log`
  - output: `/hot/software/pipeline/pipeline-recalibrate-BAM/Nextflow/development/1.0.2/sfitz-add-tool-choice/A-mini-n2-recal-only`
- NFTest - default
  - log:`/hot/software/pipeline/pipeline-recalibrate-BAM/Nextflow/development/1.0.2/sfitz-add-tool-choice/slurm-339411.out`
  - output: `/hot/software/pipeline/pipeline-recalibrate-BAM/Nextflow/development/1.0.2/sfitz-add-tool-choice/A-mini-n2`
  - cases:    default set 

**older logs below, with potential sample ID issue**
- Indel Realignment Only
  - sample:    A-mini
  - input yaml: /hot/code/sfitzgibbon/GitHub/uclahs-cds/pipeline-recalibrate-BAM/development-extras/yaml/amin.yaml
  - config:    /hot/code/sfitzgibbon/GitHub/uclahs-cds/pipeline-recalibrate-BAM/development-extras/config/indel_realignment_only.config
  - output:    /hot/software/pipeline/pipeline-recalibrate-BAM/Nextflow/development/1.0.2/sfitz-add-tool-choice/amin-indel_realignment_only
  - log: /hot/code/sfitzgibbon/GitHub/uclahs-cds/pipeline-recalibrate-BAM/development-extras/amin-logs/slurm-336566.out

- Recalibration Only
  - sample:    A-mini
  - input yaml: /hot/code/sfitzgibbon/GitHub/uclahs-cds/pipeline-recalibrate-BAM/development-extras/yaml/amin.yaml
  - config:    /hot/code/sfitzgibbon/GitHub/uclahs-cds/pipeline-recalibrate-BAM/development-extras/config/recalibrate_only.config
  - output:    /hot/software/pipeline/pipeline-recalibrate-BAM/Nextflow/development/1.0.2/sfitz-add-tool-choice/amin-recalibrate_only
  - log: /hot/code/sfitzgibbon/GitHub/uclahs-cds/pipeline-recalibrate-BAM/development-extras/amin-logs/slurm-336565.out

# Checklist
<!-- Please read each of the following items and confirm by replacing the [ ] with a [X] -->

- [x] I have read the [code review guidelines](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3187646/Code+Review+Guidelines) and the [code review best practice on GitHub check-list](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3189956/Code+Review+Best+Practice+on+GitHub+-+Check+List).

- [x] I have reviewed the [Nextflow pipeline standards](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3193890/Nextflow+pipeline+standardization).

- [x] The name of the branch is meaningful and well formatted following the [standards](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3189956/Code+Review+Best+Practice+on+GitHub+-+Check+List), using \[AD\_username (or 5 letters of AD if AD is too long)]-\[brief\_description\_of\_branch\].

- [x] I have set up or verified the branch protection rule following the [github standards](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3190380/GitHub+Standards#GitHubStandards-Branchprotectionrule) before opening this pull request.

- [x] I have added my name to the contributors listings in the ``manifest`` block in the `nextflow.config` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

- [x] I have added the changes included in this pull request to the `CHANGELOG.md` under the next release version or unreleased, and updated the date.

- [x] I have updated the version number in the `metadata.yaml` and `manifest` block of the `nextflow.config` file following [semver](https://semver.org/), or the version number has already been updated. (*Leave it unchecked if you are unsure about new version number and discuss it with the infrastructure team in this PR.*)

- [x] I have tested the pipeline using NFTest, _or_ I have justified why I did not need to run NFTest above.

sorelfitzgibbon committed:
add tool choice

sorelfitzgibbon committed:
add F8 config

sorelfitzgibbon committed:
simplify readme

sorelfitzgibbon committed:
update changelog

yashpatel6 commented on main.nf:
```diff
@@ -146,46 +159,123 @@ workflow {
+            .set{ samples_for_merge }
+            
+    } else if (!params.run_indel_realignment && params.run_base_recalibration) {
+        // Base recalibration only on raw BAMs
+        // Create interval-based entries to enable ApplyBQSR parallelization
+        // BaseRecalibrator will deduplicate identical BAMs automatically
+        validated_samples_with_index
+            .combine(input_ch_intervals)
+            .map{ sample, interval ->
+                [
```
**comment**: I'm not a fan of passing messy inputs to a workflow and expecting downstream processing to clean it up; I would suggest properly handling the inputs here

yashpatel6 reviewed (commented):
The approach used here introduces a lot of duplicated code. Implementation should be structured such that the choice of IndelRealignment/BQSR affects which workflows/processes are called by manipulating the connecting channels rather than brute-force coding all possible choices with large duplications

sorelfitzgibbon committed:
fix loss of sample.id

sorelfitzgibbon committed:
parallelize run_BaseRecalibrator

sorelfitzgibbon committed:
Fix sporadic sample ID issue in BaseRecalibrator

- Preserve individual sample IDs through indel realignment workflow
- Use sample_id instead of params.patient_id for filename generation
- Replace filename parsing with direct sample_id usage
- Prevents sporadic BaseRecalibrator .grp files having patient_id instead of sample_id

This fixes the issue where all interval outputs would sometimes be the same file
due to incorrect sample ID extraction from BAM filenames.

sorelfitzgibbon committed:
add testfile

sorelfitzgibbon commented:
/fix-tests

sorelfitzgibbon committed:
Autofix Nextflow configuration regression tests

sorelfitzgibbon committed:
remove tmp skip QC

yashpatel6 reviewed (commented):
Another issue: the tool choice implementation is resulting in some processes such as IndelRealignment ignoring one of the samples when multiple samples are given.

Is there also any testing being done to verify that the output BAMs are identical to the ones produced by the current version of recalibrate-BAM?

sorelfitzgibbon commented:
> Another issue: the tool choice implementation is resulting in some processes such as IndelRealignment ignoring one of the samples when multiple samples are given.

Ahhh great!!  This explains some minor differences I was seeing in the PCAWG BAMs.  I had forgotten that we want to indel realign all samples together.  Fixing now and will update NFtesting.  I'll redo testing on large BAMs on the [most recent branch](https://github.com/uclahs-cds/pipeline-recalibrate-BAM/pull/97) after bringing in this fix.

> Is there also any testing being done to verify that the output BAMs are identical to the ones produced by the current version of recalibrate-BAM?

Sorry, I had put this in the next PR:  Comparing to expected for the BAMs fails for NFTest but only due to the headers being different.   The a-mini BAMs aren't big enough to see the subtle difference of indel realigning together.  As I said above, there were minor differences in the larger BAMs which will probably be explained by the separated indel realignment.
```
samtools view /hot/software/pipeline/pipeline-recalibrate-BAM/Nextflow/development/unreleased/sfitz-add-tool-choice/A-mini-n2/recalibrate-BAM-1.0.2/TWGSAMIN000001/GATK-4.2.4.1/output/BWA-MEM2-2.2.1_GATK-4.2.4.1_A-mini_S2-v1.1.5.bam > tmp1
samtools view /hot/software/pipeline/pipeline-recalibrate-BAM/Nextflow/development/output/BWA-MEM2-2.2.1_GATK-4.2.4.1_A-mini_S2-v1.1.5.bam > tmp2
diff tmp1 tmp2 | head -n 30
```

nwiltsie commented:
> Comparing to expected for the BAMs fails for NFTest but only due to the headers being different.

I've got a [custom comparison script](https://github.com/uclahs-cds/pipeline-call-GeneticAncestry/blob/ea4a4a4271f94cdab24c86c5f6dd7e15bfe7430d/test/bcftoolsless_vcf_comparison.py) in call-GeneticAncestry that masks out `##bcftools_.*Command` header lines - we should incorporate similar logic here.

```python
#!/usr/bin/env python3
"""A simple script to check equality of two VCFs while ignoring any embedded BCFtools commands."""

import argparse
import gzip
import re
import sys


def differences_exist(actual: str, expect: str) -> bool:
    """Return True if the VCFs have differences (ignoring BCFtools commands)."""
    regex = re.compile(rb"^(##bcftools_(?:\w+?)Command=)(.+)$")
    repl = rb"\1masked"

    with gzip.open(actual, "rb") as actual_fileobj:
        with gzip.open(expect, "rb") as expect_fileobj:
            for line_number, (line_1, line_2) in enumerate(
                zip(actual_fileobj, expect_fileobj), start=1
            ):
                if line_1 != line_2:
                    clean_1 = regex.sub(repl, line_1)
                    clean_2 = regex.sub(repl, line_2)
                    if clean_1 != clean_2:
                        print(f"Mismatch on line {line_number}!")
                        print("Actual:", line_1.decode("utf-8").rstrip("\n"))
                        print("Expect:", line_2.decode("utf-8").rstrip("\n"))
                        return True

    print("No differences found")
    return False


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("actual")
    parser.add_argument("expect")

    args = parser.parse_args()

    sys.exit(int(differences_exist(args.actual, args.expect)))
```

sorelfitzgibbon committed:
fix patient level indel realignment - broken

sorelfitzgibbon commented:
Superseded by https://github.com/uclahs-cds/pipeline-recalibrate-BAM/pull/100 and https://github.com/uclahs-cds/pipeline-recalibrate-BAM/pull/101
